synapse/storage/databases/main/events_worker.py

```diff
@@ -1626,3 +1626,104 @@ def _cleanup_old_transaction_ids_txn(txn):
             "_cleanup_old_transaction_ids",
             _cleanup_old_transaction_ids_txn,
         )
+
+    async def check_if_event_is_extremity(self, room_id: str, event_id: str):
+        def check_if_event_is_extremity_txn(txn) -> bool:
+            backward_extremity_query = """
+                SELECT event_id FROM event_backward_extremities
+                WHERE room_id = ? AND event_id = ?
+                LIMIT 1
+            """
+
+            forward_extremity_query = """
+                SELECT event_id FROM event_forward_extremities
+                WHERE room_id = ? AND event_id = ?
+                LIMIT 1
+            """
+
+            all_backward_extremity_query = """
+                SELECT event_id FROM event_backward_extremities
+                WHERE room_id = ?
+                LIMIT 100
+            """
+
+            all_forward_extremity_query = """
+                SELECT event_id FROM event_forward_extremities
+                WHERE room_id = ?
+                LIMIT 100
+            """
+
+            txn.execute(all_backward_extremity_query, (room_id,))
+            all_backward_extremties = txn.fetchall()
+            txn.execute(all_forward_extremity_query, (room_id,))
+            all_forward_extremties = txn.fetchall()
+            logger.info(
+                "all_forward_extremties=%s all_backward_extremties=%s",
+                all_forward_extremties,
+                all_backward_extremties,
+            )
```

Comment:
When federated `hs2` joins, the room looks like the following for them. Should `$gHGXLXGcXfFlXdlelxOOYRw3KGRTQGvJxJKyVMotJ4w` also be a forward extremity? It's not referenced by any other `prev_events` on `hs2` yet. (@matrix-org/synapse-core any help here)

I understand that forward extremities are used for the next new event comes in as the `prev_events` and kinda wouldn't make sense to tie in `$gHGXLXGcXfFlXdlelxOOYRw3KGRTQGvJxJKyVMotJ4w` that way so there is a gap in my understanding and probably the description from https://github.com/matrix-org/synapse/blob/develop/docs/development/room-dag-concepts.md#forward-extremity

```
all_forward_extremeties=[('$AEhX_px4K4MGZ7fEtngZraN7KlK8h6AGEH5JRInI0to',)]
all_backward_extremeties=[('$qWzrvnib2Oh_XaSOVsERgWSICeR-zAqx4XkyJCBNJcA',)]
```
```
(oldest)
$ETDJdE8w0fhLCgbWOf0tMwaxTi7TCTT_CXC_XhpjqNI (1637131874827) - m.room.create
$DlVohXBO6Ts7pEOPxJ_CNlw95nYOhVfWPxuypxgsDRk (1637131874885) - m.room.member
$QliEnJH5dka-EBuHQ_9UDMmtuXtGlPcTDzHhHwRpKz0 (1637131874919) - m.room.power_levels
$kwejVaW0F90hdqOtohi5faGBj4qCy4LaJ7kjaRZQIeY (1637131874960) - m.room.join_rules
$gHGXLXGcXfFlXdlelxOOYRw3KGRTQGvJxJKyVMotJ4w (1637131874992) - m.room.history_visibility
<gap>
$AEhX_px4K4MGZ7fEtngZraN7KlK8h6AGEH5JRInI0to (1637131875789) - m.room.member
(newest)
```

---

The actual full picture on `hs1`, looks like:
```
(oldest)
$ETDJdE8w0fhLCgbWOf0tMwaxTi7TCTT_CXC_XhpjqNI (1637131874827) - m.room.create
$DlVohXBO6Ts7pEOPxJ_CNlw95nYOhVfWPxuypxgsDRk (1637131874885) - m.room.member
$QliEnJH5dka-EBuHQ_9UDMmtuXtGlPcTDzHhHwRpKz0 (1637131874919) - m.room.power_levels
$kwejVaW0F90hdqOtohi5faGBj4qCy4LaJ7kjaRZQIeY (1637131874960) - m.room.join_rules
$gHGXLXGcXfFlXdlelxOOYRw3KGRTQGvJxJKyVMotJ4w (1637131874992) - m.room.history_visibility
$5Npt2_DwUfzfXWbDYy2HMlgDPCHWa2esUn5F9dQecnk (1637131875032) - m.room.message
$qWzrvnib2Oh_XaSOVsERgWSICeR-zAqx4XkyJCBNJcA (1637131875091) - m.room.message
$AEhX_px4K4MGZ7fEtngZraN7KlK8h6AGEH5JRInI0to (1637131875789) - m.room.member
(newest)
```

Reply:
> `event_forward_extremities` can't be from `backfilled` events or `outliers`

After asking in #synapse-devs, it turns out this is by design. I think it's best to rename the method, otherwise we'll end up with multiple conflicting concepts of a forward extremity in the code.

Reply:
Thanks @squahtx, updated to `event_next_to_gap` and the new focus on just gaps